internal/provisioner/plain/controllers/bundleinstance_controller.go

```diff
@@ -103,12 +103,13 @@ func (r *BundleInstanceReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 
 	b := &rukpakv1alpha1.Bundle{}
 	if err := r.Get(ctx, types.NamespacedName{Name: bi.Spec.BundleName}, b); err != nil {
-		if apierrors.IsNotFound(err) {
-			bi.Status.InstalledBundleName = ""
+		currentInstalledCondition := meta.FindStatusCondition(bi.Status.Conditions, "Installed")
+		if apierrors.IsNotFound(err) && currentInstalledCondition != nil && currentInstalledCondition.Status != "True" {
+			bi.Status.InstalledBundleName += ""
```

Comment:
I'm not sure I'm following this code. It looks like a no-op?

Its possible to have both a successfully installed bundle (from a successful previous reconciliation) and an invalid bundle during this reconciliation.

Consider a scenario where the BI first has a valid bundle "foo-v1", and then is changed to a non-existent bundle "foo-v2". I'd expect to see this in the conditions:
```yaml
status:
  conditions:
  - type: Installed
    status: "True"
    reason: InstallSucceeded
  - type: HasValidBundle
    status: "False"
    reason: bundle "foo-v2" not found
  installedBundle: foo-v1
```

Before we return from the reconcile method we need to do two different things:
1. Look to see if anything is currently installed, and set the `Installed` condition accordingly (an error looking this up should result in an `Unknown` status).
2. Lookup and validate the bundle, and set the `HasValidBundle` condition accordingly (an error other than "not found" when looking up the bundle should probably result in an `Unknown` status. Any further validation errors should result in a `False` status).

Only after we've initialized both conditions should we consider ending the reconciliation by returning.

Reply:
Ah, I see what you are saying here. That's a really good observation. Let me iron that out.

Reply:
@joelanford I believe that I've covered this now. Would you mind giving it a look?